src/contract.ts

```diff
@@ -166,74 +149,131 @@ function serializeArgs(
     }
 }
 
+export class ContractMethod {
+    #contract: Contract;
+    public get contract(): Contract {
+        return this.#contract;
+    }
+    #arguments: any[];
+    public get arguments(): any[] {
+        return this.#arguments;
+    }
+    #method: AbiFunction;
+    public get method(): AbiFunction {
+        return this.#method;
+    }
+
+    call?(
+        wallet: Wallet,
+        opts?: FunctionCallOptions
+    ): Promise<void | FinalExecutionOutcome>;
+
+    view?(): Promise<any>;
+
+    /**
+     * @param contract NEAR Contract object
+     * @param fn ABI function object
+     * @param args Arguments to pass to the function
+     */
+    constructor(contract: Contract, fn: AbiFunction, args: any[]) {
+        [this.#method, this.#arguments, this.#contract] = [fn, args, contract];
+        if (fn.is_view) {
+            this.view = async (): Promise<any> => {
+                const returnBytes = await viewInternal(
+                    contract.connection,
+                    contract.contractId,
+                    fn.name,
+                    serializeArgs(fn.name, args, fn.params)
+                );
+                // TODO deserialize based on protocol from schema
+                return deserializeJSON(returnBytes);
+            };
+            Object.defineProperty(this.view, 'name', {
+                writable: false,
+                value: `ContractMethod[${fn.name}].view`,
+            });
+        } else {
+            this.call = async (
+                account,
+                opts
+            ): Promise<void | FinalExecutionOutcome> => {
+                // Using inner NAJ APIs for result for consistency, but this might
+                // not be ideal API.
+                return callInternal(
+                    account,
+                    contract.contractId,
+                    fn.name,
+                    serializeArgs(fn.name, args, fn.params),
+                    opts
+                );
+            };
+            Object.defineProperty(this.view, 'name', {
+                writable: false,
+                value: `ContractMethod[${fn.name}].call`,
+            });
+        }
+    }
+}
+
+export class ContractMethods {
+    readonly [fn: string]: (...args: any[]) => ContractMethod;
+
+    /**
+     * @param contract NEAR Contract object
+     */
+    constructor(contract: Contract) {
+        if (!contract.abi)
+            throw new Error("Can't create ContractMethods without ABI");
+        // Create method on this contract object to be able to call methods.
+        for (const fn of contract.abi.body.functions) {
+            const handler = (...args): ContractMethod => {
+                return new ContractMethod(contract, fn, args);
```

Comment:
Any reason to create a new instance on every invocation?

Reply:
Could probably be more succinctly named, but this is effectively an object that that associates arguments to a method.

```js
> let add_method: ContractMethod = contract.methods.add([1, 2], [3, 4])

> add_method.arguments
[[1, 2], [3, 4]]

> await add_method.view()
[4, 6]
```

Got any ideas for an alternative name?

Reply:
Ah, I see, missed the `args` part.

Maybe `ContractMethodInvocation`? A bit wordy though

Reply:
Yeah, that works, if needed, we can revise the name later. Thanks!